### PR TITLE
Share attribute required feature

### PR DESCRIPTION
**Pull request by mrow4a**
Related Issue: https://github.com/owncloud/core/pull/33994#issuecomment-481226241

When adding extra share attributes, the ability to specify not only incompatible permissions (`incompatibleAttributes`), but also mandatory permissions is required (`requiredPermissions`).

@Galegod @LinneyS

### CODE REVIEW

**Comment by mrow4a** [comment]
@LinneyS can you sing the CLA? If you want to have it merged, we need it before RC

**Comment by LinneyS** [comment]
We still expect these edits to be applied.
https://github.com/ONLYOFFICE/onlyoffice-owncloud/blob/184ea571612fcc770bffb1124d92014676179ca5/js/main.js#L331

**Comment by mrow4a** [comment]
@PVince81 lets wait with RC for this - I am in good progress with javascript tests. It would be good to have full feature set for both Collabora and OnlyOffice for 10.2 (@pmaier1 would also agree I think)

**Comment by mrow4a** [comment]
@PVince81 thanks! Suggested changes in OnlyOffice to work with 10.2 are applied - https://github.com/ONLYOFFICE/onlyoffice-owncloud/pull/255

**Comment by mrow4a** [comment]
@PVince81 finished, please review

**Comment by LinneyS** [comment]
Hi @mrow4a 
What do you think about this:
ownCloud

When I mark **b**, **c** is hidden
When I mark **a**, **b** is hidden and **c** remains.

When I mark **b**, and then mark **a**, **b** is hidden, but **c** remains hidden.

<details>
<summary>Code used</summary>
<p>

```javascript
model.registerShareAttribute({
    "scope": "s",
    "key": "a",
    "default": false,
    "label": "a"
});
model.registerShareAttribute({
    "scope": "s",
    "key": "b",
    "default": false,
    "label": "b",
    "incompatibleAttributes": [
        {
            "scope": "s",
            "key": "a",
            "enabled": true
        }
    ]
});
model.registerShareAttribute({
    "scope": "s",
    "key": "c",
    "default": false,
    "label": "c",
    "incompatibleAttributes": [
        {
            "scope": "s",
            "key": "b",
            "enabled": true
        }
    ]
});
```
</p>
</details>

**Comment by mrow4a** [comment]
@LinneyS I am not sure if this is related to this PR. We could move it to OnlyOffice issue. @LinneyS is the change in this PR allowing you add your checkboxes logic?

File: core/js/tests/specs/sharedialogshareelistview.js [hunk 1]
```diff
@@ -222,6 +256,139 @@ describe('OC.Share.ShareDialogShareeListView', function () {
 			expect(listView.$el.find("input[name='mailNotification']").hasClass('hidden')).toEqual(false);
 		});
 
+		it('share attribute can be unchecked', function () {
```
Comment (PVince81) [inline core/js/tests/specs/sharedialogshareelistview.js]: note, usually test scenarios are in the form "it XXX" and here "it share attribute ..." isn't a proper sentence.

example is: "it unchecks the attribute when clicked"

File: core/js/tests/specs/shareitemmodelSpec.js [hunk 2]
```diff
@@ -589,6 +593,328 @@ describe('OC.Share.ShareItemModel', function() {
...
+		 * request (request send to the server)
+		 *
+		 * @return {OC.Share.Types.ShareAttribute[]}
+		 */
+		function parseLastRequestAttributes() {
+			var requestBody = OC.parseQueryString(fakeServer.requests[0].requestBody);
+
+			var i = -1;
+			var attributes = [];
+			_.map(Object.keys(requestBody), function(key) {
```
Comment (PVince81) [inline core/js/tests/specs/shareitemmodelSpec.js]: I didn't check whether all modern browsers support `Object.keys()`. to be safe you can use `_.keys()` instead in all locations

**Review by PVince81 — APPROVED** [review]
:+1: looks fine to me.

you can optionally address my comments before merging

**Comment by PVince81** [comment]
after merging let's backport to release-10.2.0.

considering that the original thing was new APIs, we should have those released in their final form

**Comment by PVince81** [comment]
Fixes https://github.com/owncloud/core/issues/34975

File: core/js/tests/specs/shareitemmodelSpec.js (hunk 2 above)
Comment (mrow4a) [reply core/js/tests/specs/shareitemmodelSpec.js]: Ok, will check!

**Comment by PVince81** [comment]
moving forward, skipping optional items

**Comment by mrow4a** [comment]
@PVince81 some of the fixes to this PR are addressed here https://github.com/owncloud/core/pull/35111

**Comment by phil-davis** [comment]
This was backported to `release-10.2.0` in PR #35095
The release branch changes were merged into `stable10` after release.

**Pull request merged**